### PR TITLE
chore: bump Node.js nvm version, update package engine definitions

### PR DESCRIPTION
**Pull request by rhyslbw**
# Context
We’re currently on the old LTS, so should bump to the latest (v18) and ensure our package.json engine definitions are consistent and cover all versions up to the next interim major (v19)

# Proposed Solution
Bump to the latest LTS version for development, shift the lower engine bound in each package and root to the latest from v14, and remove the upper bound. Update workflows to use the same v18 version.

### CODE REVIEW

File: packages/e2e/package.json [hunk 1]
```diff
@@ -100,7 +103,6 @@
     "@types/bunyan": "^1.8.8",
     "@types/jest": "^28.1.2",
     "@types/lodash": "^4.14.182",
-    "@types/node": "14.18.12",
```
Comment (mirceahasegan) [inline packages/e2e/package.json:103]: are these being picked up by the root package.json?

**Review by mirceahasegan — APPROVED** [review]

**Review by DominikGuzei — APPROVED** [review]

**Review by IvayloAndonov — DISMISSED** [review]

File: packages/e2e/package.json (hunk 1 above)
Comment (rhyslbw) [reply packages/e2e/package.json:103]: Yes, but I now may need to revert this change and just duplicate it in every package, as I think there's a dependency nominating it as a peer dep. Will also look into the CI failure, which isn't failing locally

**Comment by rhyslbw** [comment]
CI failure resolved in https://github.com/input-output-hk/cardano-js-sdk/pull/499/commits/f28f355e2a67e593c9dd38eefb3d6c6196217a54

It was a breaking change introduced in https://github.com/nodejs/node/pull/39987

**TLDR**; The host DNS preferences are now respected, so asserting the response is `127.0.0.1` will fail on systems that prefer IPv6.

https://stackoverflow.com/questions/72648487/node-js-16-17-changed-resolution-of-localhost

File: packages/e2e/package.json (hunk 1 above)
Comment (rhyslbw) [reply packages/e2e/package.json:103]: >  I think there's a dependency nominating it as a peer dep

I found a solution to this: https://github.com/input-output-hk/cardano-js-sdk/pull/499/commits/7e02986d4156431b4d63ee2748d10863d1482359

**Comment by rhyslbw** [comment]
Added vertical Node.js testing within the nightly workflow in 25d31413e84cae362d4b5f97cd267ca326625751

**Review by mkazlauskas — DISMISSED** [review]
Looks good! :rocket:

File: .github/workflows/nightly.yaml [hunk 2]
```diff
@@ -11,10 +11,11 @@ jobs:
     strategy:
       matrix:
         os: [ ubuntu-20.04 ]
+        nodejs: [ 14, 16, 18.12.0 ]
```
Comment (mkazlauskas) [inline .github/workflows/nightly.yaml:14]: This [seems to be supported](https://github.com/actions/setup-node#supported-version-syntax)
```suggestion
        nodejs: [ 14, 16, lts/*, latest ]
```

Comment (rhyslbw) [reply .github/workflows/nightly.yaml:14]: Yes that does make sense given `engines` has no upper bound. Good suggestion! I'll include it when squashing

**Review by mkazlauskas — APPROVED** [review]

**Review by mirceahasegan — DISMISSED** [review]

**Review by mirceahasegan — APPROVED** [review]
Nice work @rhyslbw  🚀
